src/toil/lib/aws/iam.py

```diff
@@ -0,0 +1,140 @@
+
+
+logger = logging.getLogger(__name__)
+
+_CLUSTER_LAUNCHING_PERMISSIONS = {"iam:CreateRole",
+                                  "iam:CreateInstanceProfile",
+                                  "iam:TagInstanceProfile",
+                                  "iam:DeleteRole",
+                                  "iam:DeleteRoleProfile",
+                                  "iam:ListAttatchedRolePolicies",
+                                  "iam:ListPolicies",
+                                  "iam:ListRoleTags",
+                                  "iam:PutRolePolicy",
+                                  "iam:RemoveRoleFromInstanceProfile",
+                                  "iam:TagRole"
+                                  }
+
+
+def check_policy_warnings(allowed_actions: Dict[str, List[str]] = {'*': []}, launching_perms: Set[str] = _CLUSTER_LAUNCHING_PERMISSIONS) -> None:
+    """
+    Check whether necessary permissions are permitted for AWS
+
+    :param allowed_actions: Dictionary containing actions allowed by resource
+    :param launching_perms: Set of required actions to launch a cluster on AWS
+    """
+    permissions = [x for x in launching_perms if helper_permission_check(x, allowed_actions["*"])]
+
+    if not launching_perms.issubset(set(permissions)):
+        raise RuntimeError("Missing permissions", permissions)
+
+    return None
+
+
+def helper_permission_check(perm: str, list_perms: List[str]) -> bool:
+    """
+    Takes a permission and checks whether it's allowed against a list of allowed permissions
+
+    :param perm: Permission to check in string form
+    :param list_perms: Permission list to check against
+    """
+    flag = False
+    for allowed in list_perms:
+        if allowed[0] == "*":
+            if perm.endswith(allowed[1:]):
+                flag = True
+
+        if allowed[0] == "*" and allowed[-1] == "*":
+            if allowed[1:-1] in perm:
+                flag = True
+
+        if allowed[-1] == "*":
+            if perm.startswith(allowed[:-1]):
+                flag = True
+
+        if allowed == perm:
+            flag = True
+    if not flag:
+
+        return False
+    else:
+        return True
+
+
+
+def test_dummy_perms() -> bool:
+    """
+    Test for success of check policy warning against dummy permissions
+    """
+    launch_tester = {'*': ['ec2:*', 'iam:*', 's3:*', 'sdb:*']}
+
+    check_policy_warnings(allowed_actions=launch_tester)
+    print("Success")
+    return True
+
+
+def get_allowed_actions() -> Dict[str, List[str]]:
+    """
+    Returns a list of all allowed actions in a dictionary which is keyed by resource permissions
+    are allowed upon.
+    """
+    aws = AWSConnectionManager()
+
+    region = zone_to_region(get_best_aws_zone() or "us-west-2a" )
+
+
+    iam: IAMClient = cast(IAMClient, aws.client(region, 'iam'))
+
+    response = iam.get_instance_profile(InstanceProfileName="fakename_toil")
+
+    role_name = response['InstanceProfile']['Roles'][0]['RoleName']
```

Comment:
Is this somehow going to be the caller's role? Why?

Reply:
During one of my pair programming sessions with Lon we were testing ways to get associated permissions, and determined we couldn't do it through the user without asking for more parameters because to get the permissions directly attached to a user we need a username and a policy name so instead we got the role associated with the instance profile and check the permissions on the role

Reply:
This produces a role name even if `fakename_toil` doesn't exist as an instance profile, right? Some kind of default role? And then you are checking the permissions on that role.

But it doesn't seem to me that the permissions on that default role are particularly likely to also be the permissions that `toil launch-cluster` has when it is running. Do you have a justification for thinking that that I don't know about? (Maybe when you run this from an EC2 machine that default role is always the same role as the current machine?) Or are we just checking the permissions of some random role instead of what we are supposed to accomplish, which is checking the permissions that are currently granted to us?